rslib/src/scheduler/states/load_balancer.rs

```diff
@@ -0,0 +1,136 @@
+// Copyright: Ankitects Pty Ltd and contributors
+// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
+
+use std::cmp::Ordering;
+use std::collections::HashSet;
+
+use crate::decks::DeckId;
+use crate::notes::NoteId;
+use crate::storage::SqliteStorage;
+
+const MAX_LOAD_BALANCE_INTERVAL: u32 = 90;
+const PERCENT_BEFORE: f32 = 0.1;
+const PERCENT_AFTER: f32 = 0.1;
+const DAYS_MIN_BEFORE: i32 = 1;
+const DAYS_MIN_AFTER: i32 = 1;
+const DAYS_MAX_BEFORE: i32 = 6;
+const DAYS_MAX_AFTER: i32 = 4;
+
+pub struct LoadBalancer<'a> {
+    today: u32,
+    note_id: NoteId,
+    deck_id: Option<DeckId>,
+    avoid_siblings: bool,
+    storage: &'a SqliteStorage,
+}
+
+impl<'a> LoadBalancer<'a> {
+    pub fn new_from_collection(
+        today: u32,
+        storage: &'a SqliteStorage,
+        note_id: NoteId,
+        avoid_siblings: bool,
+    ) -> LoadBalancer<'a> {
+        LoadBalancer {
+            today,
+            note_id,
+            avoid_siblings,
+            storage,
+            deck_id: None,
+        }
+    }
+
+    pub fn new_from_deck(
+        today: u32,
+        storage: &'a SqliteStorage,
+        note_id: NoteId,
+        deck_id: DeckId,
+        avoid_siblings: bool,
+    ) -> LoadBalancer<'a> {
+        LoadBalancer {
+            today,
+            note_id,
+            avoid_siblings,
+            storage,
+            deck_id: Some(deck_id),
+        }
+    }
+
+    pub fn find_interval(&self, interval: u32) -> u32 {
+        // if we're sending a card far out into the future, the need to balance is low
+        if interval > MAX_LOAD_BALANCE_INTERVAL {
+            return interval;
+        }
+
+        // determine the range of days to check
+        let before_range =
+            ((interval as f32 * PERCENT_BEFORE) as i32).clamp(DAYS_MIN_BEFORE, DAYS_MAX_BEFORE);
+        let after_range =
+            ((interval as f32 * PERCENT_AFTER) as i32).clamp(DAYS_MIN_AFTER, DAYS_MAX_AFTER);
+
+        let before_days = (interval as i32 - before_range).max(1);
+        let after_days = interval as i32 + after_range + 1; // +1 to make the range inclusive of the actual value
+
+        // ok this looks weird but its a totally reasonable thing
+        // I want to be as close to the original interval as possible
+        // so this enumerates out from the center
+        // i.e. 0 -1 1 -2 2 .....
+        // for optimal load balancing, it might be preferable to
+        // just default to the earliest date? it is how the old
+        // addon used to do it...
+        let intervals_to_check = (before_days..interval as i32)
+            .map(|before| before - interval as i32)
+            .chain((interval as i32..after_days).map(|after| after - interval as i32))
+            .enumerate()
+            .collect::<Vec<_>>();
```

Comment:
If the original interval is very long (e.g., 10 years), the load balancer would be slow.

Reply:
Never mind, I find it doesn't load balance cards with intervals over some threshold N.